dask/optimization.py

```diff
@@ -916,7 +916,7 @@ class SubgraphCallable(object):
     __slots__ = ('dsk', 'outkey', 'inkeys', 'name')
 
     def __init__(self, dsk, outkey, inkeys, name='subgraph_callable'):
-        self.dsk = dsk
+        self.dsk, _ = cull(dsk, outkey)
```

Comment:
I'd rather we didn't `cull` here. `fuse` with `fuse_subgraphs=True` will only submit graphs that are already culled, adding a `cull` step here adds extra unnecessary overhead. If culling is needed for atop fusion, then `cull` should be called there before the `SubgraphCallable` is created.

Reply:
This change was motivated by this test

```python
________________________________________________________ test_SubgraphCallable ________________________________________________________

    def test_SubgraphCallable():
        non_hashable = [1, 2, 3]
    
        dsk = {'a': (apply, add, ['in1', 2]),
               'b': (apply, partial_by_order, ['in2'],
                     {'function': func_with_kwargs, 'other': [(1, 20)], 'c': 4}),
               'c': (apply, partial_by_order, ['in2', 'in1'],
                     {'function': func_with_kwargs, 'other': [(1, 20)]}),
               'd': (inc, 'a'),
               'e': (add, 'c', 'd'),
               'f': ['a', 2, 'b', (add, 'b', (sum, non_hashable))],
               'g': (dontcall, 'in1'),
               'h': (add, (sum, 'f'), (sum, ['a', 'b']))}
    
        f = SubgraphCallable(dsk, 'h', ['in1', 'in2'], name='test')
        assert f.name == 'test'
        assert repr(f) == 'test'
    
        dsk2 = dsk.copy()
        dsk2.update({'in1': 1, 'in2': 2})
>       assert f(1, 2) == get_sync(cull(dsk2, ['h'])[0], ['h'])[0]

dask/tests/test_optimization.py:1120: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
dask/optimization.py:943: in __call__
    dict(zip(self.inkeys, args)))
dask/core.py:149: in get
    result = _execute_task(task, cache)
dask/core.py:119: in _execute_task
    return func(*args2)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

x = 1

    def dontcall(x):
>       raise ValueError("shouldn't be called")
E       ValueError: shouldn't be called

dask/tests/test_optimization.py:1093: ValueError
```

The dask array tests run fine either way.

In general though I don't think that adding cull here is that bad.  Cull is fast, and we're doing this only once per SubgraphCallable, not per task.

Reply:
> In general though I don't think that adding cull here is that bad. Cull is fast, and we're doing this only once per SubgraphCallable, not per task.

For atop fusion this is fine (single `SubgraphCallable` object called multiple times), but not for `fuse_subgraphs=True`, which must create a new `SubgraphCallable` object for every fused subgraph (effectively calling once per task).

If the goal is to satisfy the general scheduler test suite, perhaps add a flag to the `get` method to disable culling, and use that in `SubgraphCallable`?

Reply:
I could also just remove that test.  I think the operational question here is "Do we want subgraphcallable to avoid redundant work, and if so, what are we willing to pay?".  Before we were paying the cost of a somewhat complex `core.get` function (at least by my measurement).  Here we're paying the cost of `cull`.  I'd also be totally happy with an answer "No, we don't really care if `SubgraphCallable` avoids extra work, we're happy to handle this when calling it".

Reply:
Operationally I don't know of cases where we care about avoiding unrelated work in `SubgraphCallable`, other than that it's generally a nice idea if it has no other costs.

Reply:
Oops, missed that that was in the `SubgraphCallable` test (and not in our generic scheduler test). Yeah, I'd remove that part of the test.